### PR TITLE
Linear advance advances

### PR DESCRIPTION
**Pull request by tombrazier**
### Description

tl;dr: Linear advance is now smaller, generates a cleaner pulse train, runs the E stepper more quietly and, reportedly, no longer generates causes TMC2208 stepper drivers to go to sleep.

This PR changes how E stepper pulses are generated when `LIN_ADVANCE` is enabled. In the original implementation, the normal `pulse_phase_isr()` would generate E steps at the usual rate whilst `advance_isr()` generated E steps at a rate related to acceleration. (Note: the actual pulses to the E stepper were all generated inside `advance_isr()` which incorporated the steps generated by `pulse_phase_isr()`). This resulted in two overlapping E stepper pulse trains. This contributed to E stepper noise and caused some TMC stepper drivers (e.g. TMC2208, TMC2225) to give up and go to sleep.

Now all steps and pulses are generated inside advance_isr() as a single pulse train. (Although `pulse_phase_isr()` does still record E steps that would have been generated without linear advance so that the logic can keep track of how many steps were "advance steps" and how many were just what the gcode had requested.)

The new new code uses 32 bytes less RAM and 244 bytes less flash on my AVR machine ([configs.zip](https://github.com/MarlinFirmware/Marlin/files/9159921/configs.zip)):
Before:
```
RAM:   [========  ]  75.6% (used 6197 bytes from 8192 bytes)
Flash: [========  ]  83.9% (used 213078 bytes from 253952 bytes)
```
After
```
RAM:   [========  ]  75.3% (used 6165 bytes from 8192 bytes)
Flash: [========  ]  83.8% (used 212834 bytes from 253952 bytes)
```

It spends marginally more time in the stepper ISR, e.g. `G1 I1 E2 F1800` spends 225ms in the ISR before the change and 230ms after the change. However with the recent arc and planner optimisations, the planner queue remains full throughout.

The extra processing is a consequence of removing the `FORCE_INLINE` from `calc_timer_interval()` which is now called from three extra places. Cumulatively inlining adds 1k of code to flash. If the `FORCE_INLINE` is kept, then the new linear advance code runs faster (220ms) than the old code but, as I say, costs an extra 1k of flash.

The code change also opens the way to integrating `S_CURVE_ACCELERATION` with `LIN_ADVANCE`. All that is needed is for `Stepper::_eval_bezier_curve()` to calculate acceleration as well as speed. This would add a few multiplications but is not excessively expensive. Since `Stepper::_eval_bezier_curve()` is written in highly optimised assembler, this is work for another day.

In the course of implementing this, it was also found that there was an incorrect extra delay in the first step generated for each segment. This is resolved in aaf7d0e.

### Requirements

Cleaner pulse train for any E stepper, but especially beneficial.

### Benefits

TMC steppers no longer go to sleep.

### Related Issues

#16076
#22547

### CODE REVIEW

**Comment by Farva42** [comment]
I've tested these changes on a Creality 4.2.2 board with TMC2208's in standalone, no issues with the extruder stepper shutting off like with the current implementation where the extruder will shutdown within 30 seconds of starting a print. It also sounds less harsh at higher K values then the current implementation but that might be a little subjective.

The only downside I've found is I've needed to re-tune flow rate. my prior setup before implementing the changes was at 1 now it's around .92. 

I've also tested this on a Creality 2.2 board with A4988 drivers with no issues beside the mentioned flow adjustments.

I have only been able to test the below commits. I'll try and test the rest of the commits to see if they flow issue is resolved 
[820153b](https://github.com/MarlinFirmware/Marlin/pull/24533/commits/820153b6c57ae031fe928d72a673b7d99d542878)
[092b84a](https://github.com/MarlinFirmware/Marlin/pull/24533/commits/092b84a3ed1ee48104135b86faa8bd9e6603ffd0)
[de35fbd](https://github.com/MarlinFirmware/Marlin/pull/24533/commits/de35fbd6de72d32694f80b99abc2b3840233087a)
[4f3c183](https://github.com/MarlinFirmware/Marlin/pull/24533/commits/4f3c183fba2c6272a0420f173dbf0cf0331031ba)
[0242cfb](https://github.com/MarlinFirmware/Marlin/pull/24533/commits/0242cfb3d084ed8ea12e254e7aab1e59d6441d60)
[88bb808](https://github.com/MarlinFirmware/Marlin/pull/24533/commits/88bb80839757cebb677dcb353a8214dbf05b66ba)
[aaf7d0e](https://github.com/MarlinFirmware/Marlin/pull/24533/commits/aaf7d0e9012f0caab23568b301ba150298cf1f98)

**Comment by tombrazier** [comment]
Commit a55f315 should resolve the flow issue.

**Comment by thinkyhead** [comment]
Some changes applied:
- Moved the methods to cpp file for a neater class header.
- Replaced `loops` pointer with reference, as code was originally C.
- Since `step_rate` is already compared to `min_step_rate` the logic can be optimized for the <= case.
- Some `uint16_t` are replaced with `uintptr_t` where appropriate.
- Fixed some gcode/ includes of `stepper.h` that should have been changed to `planner.h` a while ago.

**Comment by The-EG** [comment]
I'm getting some odd results. I'll admit, I've never used linear advance before, so I originally chalked this up to me not knowing what I was doing, but this does seem really odd.

First, my results from running the [calibration pattern](https://marlinfw.org/tools/lin_advance/k-factor.html). This is for PLA on my Ender 3 v2, with a stock 4.2.2 board (2208s in standalone). 0.2 step, from K=0 all the way up to k=4:
image

At first I thought maybe linear advance doesn't work with my machine, except I had accidently run a few calibration patterns earlier without the PR changes (I had forgotten to run git apply without `--check` after verifying it would apply cleanly).

This is the results I got without the PR, this time with PETG:
image

This is much more like what I expected to see! Of course, I can't test this because the driver will lock up before it can even complete the first layer. In fact, I wasn't even able to reproduce the above with PLA because it would give up part way through the calibration pattern too.

As a test, I decided to try to pick the 'best' K value from my results and try it out anyway.

With K=0 (so LA off); a little underextrusion on the top layer, but otherwise my printer/profiles are fairly tuned:
image

Based on the above, and a more detailed 0.1 step print, I decided to try K=0.5:
image
It's both under and over extruded. Also notice the stringing on the back left corner.

For S&Gs, I also tried K=1.0 in case anyone said a bowden setup should be using higher K:
image
This is even worse and is completely overextruded.

What's going on here? My results with this PR can't be right, can they??

**Comment by The-EG** [comment]
After some conversation on discord I tried a much smaller step and range for the calibration pattern. I'd say the results are similar. It looks as if the initial increase is happening to compensate for the acceleration, but the corresponding decrease for deceleration isn't 
PXL_20220727_181617011

**Comment by tombrazier** [comment]
@The-EG Just to check the facts as I think I have understood them:

- Your Ender 3 V2 has a bowden exruder and TMC2208 stepper drivers.
- Without the PR you can run the LA calibration pattern with PETG but not PLA (because the TMC2208 shuts down).
- The PETG calibration pattern without the PR looks as expected with the thin line on acceleration getting progressively thicker and the thick line on deceleration getting progressively thinner with higher K values.
- With the PR and using PLA, the acceleration part of the test behaves as expected but the deceleration part does not.
- Test prints show degraded print quality using the PR with PLA and K greater than zero.

Some questions:
- I think you are implying that the PR does indeed solve the 2208 lockup problem. Is that correct? If so, this is good news and confirms @Farva42's results. It also means it's worth getting this PR right if there is something wrong with it.
- I am surprised that the test pattern does not cause the 2208s to lock up with PETG without the PR. What is different in your gcode between PETG and PLA?
- Can you run the identical PETG test you showed above but this time with the patch applied?
- On your machine, is it possible for you to see the filament moving? Can you see a retraction in the deceleration phase?
- Ditto with seeing the E stepper moving - do you see it reverse direction during deceleration? This reversal should become increasingly pronounced with higher K values.
- Could it be that your E jerk setting is too high, resulting in missed steps or filament slipping? What do the test patterns look like with a lower E jerk?

**Comment by tombrazier** [comment]
Oh, and have you tuned your extruder for backlash?

**Comment by The-EG** [comment]
>Your Ender 3 V2 has a bowden exruder and TMC2208 stepper drivers.

Correct. 

> Without the PR you can run the LA calibration pattern with PETG but not PLA (because the TMC2208 shuts down).

Not sure if it's dependent on the material, but I haven't been able to get a calibration pattern to complete on PLA without this PR to test things out.

> The PETG calibration pattern without the PR looks as expected with the thin line on acceleration getting progressively thicker and the thick line on deceleration getting progressively thinner with higher K values.

Yep, and this is what I was expecting to see *with* the PR.

> With the PR and using PLA, the acceleration part of the test behaves as expected but the deceleration part does not.
Test prints show degraded print quality using the PR with PLA and K greater than zero.

Yes, however this behavior also occurred with PETG before that spool ran out. I did try a test print with PETG as well (~k=0.8) but it was overextruding so badly that the nozzle was dragging through the print after about 10 layers or so.
Once I ran out of PETG I decided to switch to PLA since it may be an easier material to tune 🤷‍♂️ .

> I think you are implying that the PR does indeed solve the 2208 lockup problem. Is that correct? If so, this is good news and confirms @Farva42's results. It also means it's worth getting this PR right if there is something wrong with it.

Yes, the good news from all of this is I can indeed verify that this PR fixes the 2208 lockup issue.

> I am surprised that the test pattern does not cause the 2208s to lock up with PETG without the PR. What is different in your gcode between PETG and PLA?

I'm just as confused. I tried to get a PLA test pattern to print multiple times *without* the PR. The only differences would be the temperatures and various k steps/ranges that I tried. I did try the exact same ranges as I used for the PETG print.

> Can you run the identical PETG test you showed above but this time with the patch applied?

This is one of my test patterns *with* the PR with the same PETG as the 'without' test:
image
It exhibits the exact same deceleration thing as the PLA tests.

If you'd like I can also run the 0.02 step test with a similar filament (just different color).

> On your machine, is it possible for you to see the filament moving? Can you see a retraction in the deceleration phase?
Ditto with seeing the E stepper moving - do you see it reverse direction during deceleration? This reversal should become increasingly pronounced with higher K values.

Yes. I have a basic extruder, which is more or less the stock extruder but metal. I've watched it while printing during multiple of these tests. The only time I've seen it miss steps during the test is at high K values when it was priming for the acceleration. I'll see if I can take a video.

> Could it be that your E jerk setting is too high, resulting in missed steps or filament slipping? What do the test patterns look like with a lower E jerk?

What values would you recommend? My normal EJERK is 5.0, but I did raise it to 10.0 due to the `#error`. This was the case for both the with and without PR tests. 
In any case, as stated above the only time I've seen it miss steps during the test is at high K values when it was priming for the acceleration. In that case I could *hear* it too (*thunk!*).

Also, thanks for you effort on this and for taking a look. I was really excited by this finally working on this printer and these results just really mystify me.

edit:

> Oh, and have you tuned your extruder for backlash?

Not yet, but I can look into that. Although I'd be curious why that would manifest with the PR but not without it.

**Comment by Nikolay-Po** [comment]
Looking at photo in [message above](https://github.com/MarlinFirmware/Marlin/pull/24533#issuecomment-1198576828), it seems to me, that the sign of correction is missed. After speed decreasing, extrusion pressure should decreasing but it is increasing. Just fix the sign of correction.

**Comment by tombrazier** [comment]
I have replicated the error on the simulator. There is something wrong with advance step counting, resulting in the code thinking that there is no need to reverse the advance steps during deceleration.

**Comment by The-EG** [comment]
CI tests aside, those last two commits appear to have resolved the issue I was seeing:
PXL_20220729_180533992 MP

This is the expected behavior, and a cube printed at K=0.49 looks quite good (the elephant's foot is probably due to me scraping multiple calibrations off the bed and it's starting to drift):
PXL_20220729_185824468

That's already noticeably better (look at the middle of the 'X' compared to K=0 above), and I'm sure I need to tweak some slicer settings still.

I'll continue to do more testing and some longer prints too.

**Comment by tombrazier** [comment]
Excellent! And I agree the cube looks better than with k=0. From the gaps in your top layer, I wonder about your extrusion multiplier. Have you tuned it for that filament?

CI tests failing because not all Marlin device are 3D printers. I need to put in a check for extruders existing. That will be force push so this comment will soon not make sense any more.

**Comment by The-EG** [comment]
> I wonder about your extrusion multiplier. Have you tuned it for that filament?

I have, but I suspect I may need to retune it after this and the fact that was several spools ago. And this printer has a 0.6mm nozzle on it instead of a 0.4mm. I usually print PETG with this printer so most of my PLA profiles are tuned more for a smaller nozzle (which is why I originally started in PETG). And I was monkeying with the infill overlap recently too...

In any case, that all seems to be more slicer related. I expected that I'd need to retune slicer setting once LA was working, so I'm not disappointed there. This seems to be working well now for my config, but I'll chime in if anything else odd pops up.

**Comment by tombrazier** [comment]
Thanks so much for the testing. I am really glad you caught that bug.

**Comment by XDA-Bam** [comment]
It's been a while since this has been merged, but I just wanted to report that this PR had a big impact on the E-stepper behaviour for my printer. It's running much smoother and quieter now. Good work!

**Comment by tombrazier** [comment]
@XDA-Bam thank you.

**Comment by x40-Community** [comment]
I have a Weedo X40 with TMC2208 driver. Unfortunately, Linear Advanced does not work.
The firmware is based on Marlin 2.0.8 and has been extended by a GUI connection from the manufacturer.
Is it possible to integrate the patch?

**Comment by Nikolay-Po** [comment]
> Is it possible to integrate the patch?

The most straightforward way is to ask the manufacturer to rebuild the FW with newer Marlin version. Have you tried that?
Also, there is a chance they will give you firmware source code, may be after NDA signing. Then you will be able (after some work) to switch Marlin version keeping propiertary UI.

**Comment by x40-Community** [comment]
> > Is it possible to integrate the patch?
> 
> The most straightforward way is to ask the manufacturer to rebuild the FW with newer Marlin version. Have you tried that? Also, there is a chance they will give you firmware source code, may be after NDA signing. Then you will be able (after some work) to switch Marlin version keeping propiertary UI.

Weedo is a small manufacturer that hasn't made any improvements for two years. Luckily we have the source code and have already been able to integrate some bugs into improvements.
All we really need is a bug fix to make Linear Advance work.
We would also pay money if someone integrated the patch Workable for us!

**Comment by Nikolay-Po** [comment]
> All we really need is a bug fix to make Linear Advance work.

May be it will be not so hard to integrate proprietary display and UI into current Marlin? Usually, proprietary user interface is working separately of Marlin and just sent G-codes to Marlin code. IMHO, as first step, it worth to configure current Marlin for your printer without a UI by creating separate pins.h and appropriate configuration and configuration_adv.h files. Then, after printer operation through serial port (USB connection) is confirmed, try to implement proprietary UI and SD card interface (if applicable), using 2.0.8 source code as a reference.
I think current 2.1.x is quite different after several movement improvements. But the last word is for patch author. I can mistaken.

**Comment by tombrazier** [comment]
I think @Nikolay-Po is on the right track. A short term backport of the LA fix might conceivably be easiest right now but if you want all the goodness of Marlin as it evolves you're eventually going to want to move the UI code and appropriate pins config into Marlin proper.

To answer the original question, I had a go at applying this PR to 2.0.8 but it's not trivial.

**Comment by thisiskeithb** [comment]
> Is it possible to integrate the patch?

<hr>

This Issue Queue is for Marlin bug reports and development-related issues, and we prefer not to handle user-support questions here. (As noted on [this page](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/.github/contributing.md#i-dont-want-to-read-this-whole-thing-i-just-have-a-question).) For best results getting help with configuration and troubleshooting, please use the following resources:

- [Marlin Documentation](https://marlinfw.org) hosted on MarlinFW.org
- [Marlin Forum](https://forums.reprap.org/list.php?415) hosted on RepRap.org
- [Marlin Firmware](https://www.facebook.com/groups/1049718498464482/) Facebook Group 
- [Marlin Firmware for 3D Printers](https://www.facebook.com/groups/3Dtechtalk/) Facebook Group
- [Marlin Configuration](https://www.youtube.com/results?search_query=marlin+configuration) on YouTube
- [Marlin Firmware](https://discord.gg/n5NJ59y) Discord server

**Comment by x40-Community** [comment]
I compared the Weedo X40 firmware (Idex printer) against the Marlin 2.0.8. A lot has changed in the Weedo firmware. I don't think the changes can easily be integrated into a newer Marlin version.
It's a pity really, then I'll probably have to deactivate Linear Advance and tick it off.

**Comment by ttyusupov** [comment]
I see related https://github.com/MarlinFirmware/Marlin/issues/16076 is applicable to Marlin 1.1.x, any plans to backport this fix into bugfix-1.1.x branch?

**Comment by thisiskeithb** [comment]
> I see related https://github.com/MarlinFirmware/Marlin/issues/16076 is applicable to Marlin 1.1.x, any plans to backport this fix into bugfix-1.1.x branch?

@ttyusupov: No. Short of making some small updates to allow some older versions of Marlin to use newer PIO / pin platform & library versions, `1.1.x` won't receive any more updates. You'll need to migrate to `2.1.x`.
